tapchannel/aux_traffic_shaper.go

```diff
@@ -172,18 +175,52 @@ func (s *AuxTrafficShaper) PaymentBandwidth(htlcBlob,
 		return 0, fmt.Errorf("error decoding HTLC blob: %w", err)
 	}
 
-	// localBalance is the total number of local asset units.
-	localBalance := cmsg.OutputSum(commitment.LocalOutputs())
+	// With the help of the latest HtlcView, let's calculate a more precise
+	// local balance. This is useful in order to not forward HTLCs that may
+	// never be settled. Other HTLCs that may also call into this method are
+	// not yet registered to the commitment, so we need to account for them
+	// manually.
+	computedLocal, decodedView, err := ComputeLocalBalance(
+		commitment, htlcView,
+	)
+	if err != nil {
+		return 0, err
+	}
 
-	// There either already is an amount set in the HTLC (which would
-	// indicate it to be a direct-channel keysend payment that just sends
-	// assets to the direct peer with no conversion), in which case we don't
-	// need an RFQ ID as we can just compare the local balance and the
-	// required HTLC amount. If there is no amount set, we need to look up
-	// the RFQ ID in the HTLC blob and use the accepted quote to determine
-	// the amount.
+	log.Tracef("Computed asset HTLC View: commitmentLocal=%v, "+
+		"computedLocal=%v, nextHeight=%v, thisHtlc=%v, newView=%v",
+		cmsg.OutputSum(commitment.LocalOutputs()), computedLocal,
+		htlcView.NextHeight, htlc.Amounts.Val.Sum(), func() string {
+			return prettyPrintLocalView(*decodedView)
+		}())
```

Comment:
Why? This removes the whole benefit of the function callback in the first place...

Reply:
Hmm... Very weird. It seems like it's not enough that it's the correct function type, it must be of type `LogClosure` (the reason for which I don't understand, must be some explicit configuration somewhere).
So can just be:
```
lnutils.NewLogClosure(func() string {
			return prettyPrintLocalView(*decodedView)
		})
```

Reply:
Ah, of course... Because the `LogClosure` type has a `String()` method:
```
// String invokes the underlying function and returns the result.
func (c LogClosure) String() string {
	return c()
}
```

Makes sense now. Sorry, gave you bad advice before then. I thought being able to pass in a function was a feature of the logging system. Turns out it was actually a clever trick.